src/OSPSuite.R/Services/ConcurrentSimulationRunner.cs

```diff
@@ -0,0 +1,74 @@
+﻿using OSPSuite.Core.Domain;
+using OSPSuite.Core.Domain.Data;
+using OSPSuite.Core.Domain.Services;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using SimulationRunOptions = OSPSuite.R.Domain.SimulationRunOptions;
+
+namespace OSPSuite.R.Services
+{
+   public interface IConcurrentSimulationRunner
+   {
+      /// <summary>
+      /// General simulation options
+      /// </summary>
+      SimulationRunOptions SimulationRunOptions { get; set; }
+
+      /// <summary>
+      /// Number of cores to use concurrently. Use 0 or a negative value for using the maximum available.
+      /// </summary>
+      int NumberOfCores { get; set; }
+
+      /// <summary>
+      /// List of Simulations
+      /// </summary>
+      void AddSimulation(IModelCoreSimulation simulation);
+
+      void Clear();
+
+      /// <summary>
+      /// Runs all preset settings concurrently
+      /// </summary>
+      /// <returns></returns>
+      SimulationResults[] RunConcurrently();
+   }
+
+   public class ConcurrentSimulationRunner : IConcurrentSimulationRunner
+   {
+      private IConcurrencyManager _concurrentManager;
+
+      public ConcurrentSimulationRunner(IConcurrencyManager concurrentManager)
+      {
+         _concurrentManager = concurrentManager;
+      }
+
+      public SimulationRunOptions SimulationRunOptions { get; set; }
+
+      public int NumberOfCores { get; set; }
+
+      private List<IModelCoreSimulation> _simulations = new List<IModelCoreSimulation>();
+
+      public void AddSimulation(IModelCoreSimulation simulation)
+      {
+         _simulations.Add(simulation);
+      }
+
+      public void Clear()
+      {
+         _simulations.Clear();
+      }
+
+      public SimulationResults[] RunConcurrently()
+      {
+         var results = _concurrentManager.RunAsync<IModelCoreSimulation, SimulationResults>(
```

Comment:
should not we create the instance of ConcurrentMAnager here? why is it injected?

Reply:
Why creating a new object every time. Also, with injection you decouple the implementations. Whoever is using the ConcurrentSimulationRunner could decide on the implementation of the ConcurrencyManager to use by just injecting it. Why should we create the object ourselves in the code and break the inversion of control?

Reply:
I did not say NEW. I say ensuring that we create a. new instance here so that there is no risk of memory leaks. Typically using a factory method to retrieve a new instance in a Dispose construct. That way, if something goes wrong, we clear after ourselves. 

So we do not create the object ourselves, we do not break inversion of control. We just ensure that we deal with memory leaks intrinsically by not holding references to objects

Reply:
I agree. Sorry, I did not understood before. I will add the changes for this.